File: orchestrator_core/render.py
from __future__ import annotations

from typing import Any, Mapping

from .model import validate_state
from .traceability import validate_execution_graph


def _items(values: list[str]) -> str:
    return "none" if not values else f"[{', '.join(values)}]"


def render_plan(state_value: Mapping[str, Any], analysis_value: Mapping[str, Any] | None = None) -> str:
    cross_check = analysis_value if analysis_value is not None and state_value.get("stages") and not state_value.get("legacy_migrated") else None
    state = validate_state(state_value, cross_check)
    analysis = validate_execution_graph(analysis_value) if analysis_value is not None else None
    lines = [
        "---", "schema_version: 1", f"state_revision: {state['state_revision']}",
        f"status: {state['status']}", f"current_stage: {state['current_stage'] or 'none'}",
        f"analysis_revision: {state['analysis_revision']}", "---", "# План реализации", "",
        "> Индекс генерируется controller; смысловые детали находятся в discovery, stage и review artifacts.", "",
        "## Состояние", "", f"- Запрос: `{state['request_id']}`",
        f"- Workflow: `{state['status']}`",
        f"- Analysis: `{state['analysis_status']}` revision {state['analysis_revision']}",
        f"- Feedback revision: {state['feedback_revision']}",
    ]
    if state["pending"]:
        lines += [
            f"- Pending: `{state['pending']['transition_id']}` / `{state['pending']['action']}`",
            f"- Pending reason: {state['pending']['reason']}",
        ]
    if state["blocker"]:
        lines += ["", "## Blocker", "", f"- Reason: {state['blocker']['reason']}", f"- Detail: {state['blocker']['detail']}"]
    lines += ["", "## Stage map", ""]
    sources = {item["id"]: item for item in analysis["stages"]} if analysis else {}
    for stage in state["stages"]:
        source = sources.get(stage["id"], {})
        lines += [
            f"### {stage['id']} — {stage['title']}",
            f"- Status: {stage['status'].upper()}", f"- Revision: {stage['revision']}",
            f"- Depends on: {_items(stage['depends_on'])}",
            f"- Affected area: {source.get('affected_area', 'unknown')}",
            f"- Primary risks: {_items(source.get('risks', []))}",
            f"- Consumes: {_items(source.get('contracts_consumed', []))}",
            f"- Produces: {_items(source.get('contracts_produced', []))}",
            f"- Details: {stage['details']}", f"- Review: {stage['review']}",
            f"- Human review: {stage['human_review']}",
            f"- Human review revision: {stage['human_revision']}",
            f"- Human review status: {stage['human_status'].upper()}",
            f"- Human review review: {stage['human_review_review']}", "",
        ]
    if analysis:
        lines += ["## Traceability", ""]
        for item in [*analysis["requirements"], *analysis["nfrs"]]:
            category = f" ({item['category']})" if "category" in item else ""
            lines.append(
                f"- `{item['id']}`{category} → `{item['stage']}` → {_items(item['scenarios'])} → {_items(item['acceptance'])}: {item['text']}"
            )
        lines.append("")
    return "\n".join(lines).rstrip() + "\n"

File: runtime/render.js
import { createHash } from "node:crypto";
import { ProtocolError } from "./schema.js";
import { newState } from "./state.js";
function yaml(value) {
    return value === null ? "null" : value;
}
export function renderPlan(state, analysis) {
    const lines = [
        "---",
        `schema_version: ${state.schema_version}`,
        `request_id: ${state.request_id}`,
        `state_revision: ${state.state_revision}`,
        `status: ${state.status}`,
        `analysis_revision: ${state.analysis_revision}`,
        `analysis_status: ${state.analysis_status}`,
        `current_stage: ${yaml(state.current_stage)}`,
        `legacy_migrated: ${state.legacy_migrated}`,
        "---",
        "",
        `# Plan: ${state.request_id}`,
        "",
        `Workflow status: **${state.status}**`,
        "",
    ];
    if (analysis) {
        lines.push("## Request", "", analysis.request.summary, "");
        if (analysis.request.outcomes.length) {
            lines.push("### Outcomes", "", ...analysis.request.outcomes.map((item) => `- ${item}`), "");
        }
    }
    lines.push("## Stage map", "");
    if (!state.stages.length)
        lines.push("No approved stage map yet.", "");
    for (const stage of state.stages) {
        lines.push(`### ${stage.id}: ${stage.title}`, "", `- Slug: ${stage.slug}`, `- Depends on: ${stage.depends_on.length ? stage.depends_on.join(", ") : "none"}`, `- Technical status: ${stage.status}`, `- Technical revision: ${stage.revision}`, `- Human status: ${stage.human_status}`, `- Human revision: ${stage.human_revision}`, `- Technical artifact: ${stage.details}`, `- Technical review: ${stage.review}`, `- Human artifact: ${stage.human_review}`, `- Human review: ${stage.human_review_review}`, "");
    }
    if (state.pending) {
        lines.push("## Pending transition", "", `- Transition: ${state.pending.transition_id}`, `- Action: ${state.pending.action}`, `- Actor: ${state.pending.actor}`, `- Stage: ${yaml(state.pending.stage)}`, `- Revision: ${state.pending.revision ?? "null"}`, `- Reason: ${state.pending.reason}`, "");
    }
    if (state.blocker)
        lines.push("## Blocker", "", `- Reason: ${state.blocker.reason}`, `- Detail: ${state.blocker.detail}`, "");
    return lines.join("\n").trimEnd() + "\n";
}
function normalizeStatus(value, fallback) {
    const normalized = value?.trim().toLowerCase().replace(/^pass(?:ed)?$/, "pass");
    return new Set(["proposed", "planning", "review", "pass"]).has(normalized ?? "") ? normalized : fallback;
}
function normalizeHumanStatus(value) {
    const normalized = value?.trim().toLowerCase().replace(/^pass(?:ed)?$/, "pass");
    return new Set(["pending", "planning", "review", "pass"]).has(normalized ?? "") ? normalized : "pending";
}
function legacyValue(block, names) {
    for (const name of names) {
        const match = block.match(new RegExp(`^\\s*(?:[-*]\\s*)?${name}\\s*:\\s*(.+?)\\s*$`, "im"));
        if (match)
            return match[1].replace(/^`|`$/g, "").trim();
    }
    return undefined;
}
function canonicalLegacyPath(value, fallback) {
    if (!value)
        return fallback;
    const normalized = value.replace(/\\\\/g, "/").replace(/^\.\//, "");
    if (normalized.startsWith("../") || normalized.startsWith("/") || /^[A-Za-z]:\//.test(normalized))
        return fallback;
    return normalized;
}
export function parseLegacySnapshot(content, requestId) {
    if (!content.trim())
        throw new ProtocolError("legacy-plan.md", "must not be empty");
    const heading = /^(?:#{2,4})\s+(?:Stage\s+)?(S\d{2})\s*(?::|[-—])?\s*(.*?)\s*$/gim;
    const matches = [...content.matchAll(heading)];
    const stages = [];
    for (const [index, match] of matches.entries()) {
        const start = match.index ?? 0;
        const end = index + 1 < matches.length ? matches[index + 1].index ?? content.length : content.length;
        const block = content.slice(start, end);
        const id = match[1];
        const title = match[2]?.trim() || `Legacy ${id}`;
        const ordinal = Number(id.slice(1));
        const rawSlug = legacyValue(block, ["Slug"]);
        const slug = rawSlug && /^[a-z0-9]+(?:-[a-z0-9]+)*$/.test(rawSlug) ? rawSlug : title.toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/^-|-$/g, "") || `legacy-${id.toLowerCase()}`;
        const revision = Math.max(0, Number.parseInt(legacyValue(block, ["Technical revision", "Revision"]) ?? "0", 10) || 0);
        const humanRevision = Math.max(0, Number.parseInt(legacyValue(block, ["Human revision"]) ?? "0", 10) || 0);
        const status = normalizeStatus(legacyValue(block, ["Technical status", "Status"]), revision > 0 ? "review" : "proposed");
        const humanStatus = normalizeHumanStatus(legacyValue(block, ["Human status"]));
        const prefix = String(ordinal).padStart(2, "0");
        const fingerprint = legacyValue(block, ["Semantic fingerprint", "Fingerprint"]);
        stages.push({
            id,
            title,
            slug,
            status,
            revision: status === "pass" ? Math.max(1, revision) : revision,
            human_status: humanStatus,
            human_revision: humanStatus === "pass" ? Math.max(1, humanRevision) : humanRevision,
            details: canonicalLegacyPath(legacyValue(block, ["Technical artifact", "Details"]), `stages/${prefix}-${slug}.md`),
            review: canonicalLegacyPath(legacyValue(block, ["Technical review", "Review"]), `reviews/${prefix}.md`),
            human_review: canonicalLegacyPath(legacyValue(block, ["Human artifact"]), `stages/${prefix}-${slug}.human-review.md`),
            human_review_review: canonicalLegacyPath(legacyValue(block, ["Human review"]), `reviews/${prefix}-human-review.md`),
            semantic_fingerprint: fingerprint && /^[0-9a-f]{64}$/i.test(fingerprint) ? fingerprint.toLowerCase() : null,
        });
    }
    return {
        schema_version: 1,
        request_id: requestId,
        source_sha256: createHash("sha256").update(content).digest("hex"),
        stages,
    };
}
export function parseLegacyPlan(content, requestId) {
    parseLegacySnapshot(content, requestId);
    const state = newState(requestId);
    state.legacy_migrated = true;
    state.status = "discovery";
    state.analysis_status = "missing";
    state.stages = [];
    state.current_stage = null;
    return state;
}
export function legacyFingerprintMatches(snapshot, analysis, stageId, fingerprint) {
    const legacy = snapshot.stages.find((stage) => stage.id === stageId);
    return Boolean(legacy?.status === "pass" && legacy.semantic_fingerprint && legacy.semantic_fingerprint === fingerprint(analysis, stageId));
}
